### DISCUSSION TITLE
Init hook for logger

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#10271] [Archive]
Hi all,

Would anyone happen to know if there is an init hook available where I could change the (pino) logger options globally?

Cheers!

**Comment by rijkvanzanten** [comment]
There's not. This file is currently creating the logger, but there's no hook included to change the loggerOptions object:

https://github.com/directus/directus/blob/8ed517708e83cece206be31347854db8be643129/api/src/logger.ts#L7-L20

This might actually be a slightly trickier one to add, as we'd have to turn the logger file into an async function in order to await the init hook. This in turn means that we'll have to touch every place we're importing the logger in, to make sure we're loading it async as well 🤔

**Reply by keesvanbemmel** [reply]
I'd be more than happy to help but also since it's touching a lot throughout Directus perhaps it's best to let you / core team handle this? Is there a way to sponsor that?

**Reply by rijkvanzanten** [reply]
In your particular use case, is this something you'd have to configure through the use of an (async) hook, or would it be sufficient to set it through Environment Variables? That would be way lighter to implement, and might already cover 95% of use cases.

**Reply by keesvanbemmel** [reply]
Well the only thing I want to achieve is to format the Pino logs to match the levels in google cloud / stackdriver logging as described here:
https://github.com/pinojs/pino/blob/master/docs/help.md

If I need to add one or more environment variables for that I'd be more than happy to!

**Reply by rijkvanzanten** [reply]
Hmmm yeah, it looks like you'll have to provide your own custom formatter function for that to work, which can't really be done through env vars (as it's JS). We could consider adding a dedicated env var for "severity flags" that's a lookup table to rename the pino levels? 🤔 

Something like:

```
LOGGER_LEVELS="trace:DEBUG,debug:DEBUG,info:INFO,warn:WARNING,error:ERROR,fatal:CRITICAL"
```

**Comment by keesvanbemmel** [comment]
I'm not entirely sure but it looks as if that serves this usecase perfectly fine!

**Reply by rijkvanzanten** [reply]
Yeah! It'd be tailor made for this use case. In addition, I think it'd be a good idea to pass on every `LOGGER_*` environment variable to the Pino configuration, just like we do for the database and storage adapters. That way you can override any of the other options as well. This logger-levels one would be a special case though, as we'd use it in a custom formatter function rather than it being used as a static value.

**Reply by keesvanbemmel** [reply]
Should be good to go at https://github.com/directus/directus/pull/10911

**Comment by JohnHillegass** [comment]
Agreed on the async rewrite complications. I had hoped to add this feature but gave up after seeing how everything is wired up. Env vars makes sense to me for this too. Another use case I'd be interested in seeing or even contributing is the ability to add a mixin/child logger everywhere the main logger is being used so default metadata can be added to the log statements. Currently we do this by doubling up on the log statements, hooking to init.response.request, creating a new logger with the attributes needed, logging then throwing the logger away. What do you think about having a LOGGER_DEFAULT_META env var which would be used to construct a child/mixin? Something like `logger.child(JSON.parse(process.env.LOGGER_DEFAULT_META))`

**Comment by jaunusa** [comment]
Maybe not the cleanest solution, but using a JS file as config ([Docs](https://docs.directus.io/configuration/config-options/#config-js)), so we could assign functions to environment variables, which are then used as pino formatters/transports?
Would also be nice to be able to change the timestamp key from "time" to something like "@timestamp".
Edit: already mentioned by @rijkvanzanten , sorry for that

**Reply by enzolupia** [reply]
Hi @jaunusa  did you give a try to this solution?

**Reply by jaunusa** [reply]
Don't think my proposal would work. I think everything assigned to process.env is converted to strings.
At the end, what is used as custom configuration for pino from env vars, needs to be offered as env var by directus. Directus internally configures pino.

**Comment by diegoleme** [comment]
If it were possible to choose a pre-configuration of log systems by env vars?

Ex.:
`LOG_PREDEFINITION: "Google Cloud Logging"`

Makes sense?

**Reply by rijkvanzanten** [reply]
That'd be cool! Though we don't really have any other presets than GCP I suppose 🤔

**Comment by noahjahn** [comment]
Without this init hook, is it possible to add a pino transport? For example, I want to add the [pino-datadog-transport](https://github.com/theogravity/pino-datadog-transport/tree/master).

**Reply by rijkvanzanten** [reply]
Pino transports generally are be able to run outside of the main process, like `directus start | pino-pretty` which means you don't need to use any custom logic or extensions to use them. I've never tried this community DD one though, so your mileage may vary

**Reply by noahjahn** [reply]
Doesn't look like that transport will work outside of the main process sadly, at least in its current state.
